### PR TITLE
Static link dependency libraries for C++ client lib and Python wrapper

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

We should try to link statically the dependencies of the C++ client library to allow an
easier distribution of binaries. 

In particular, having a self-contained python wrapper, enables user to just fetch a package from PyPI without compiling anything.

### Modifications

Changed CMake config to statically link all the dependencies.

### CODE REVIEW

**Comment by jai1** [comment]
@saandrews

File: pulsar-client-cpp/CMakeLists.txt [hunk 1]
```diff
@@ -32,24 +34,33 @@ endif(NOT LOG_CATEGORY_NAME)
...
+find_library(ZLIB_LIBRARY_PATH REQUIRED NAMES libz.a z)
+
+find_library(PROTO_LIBRARIES NAMES libprotobuf.a)
+find_library(CURL_LIBRARY_PATH NAMES libcurl.a curl)
 find_path(GTEST_INCLUDE_PATH gtest/gtest.h)
 find_path(JSON_INCLUDE_PATH jsoncpp)
-find_library(LIB_JSON jsoncpp)
+find_library(LIB_JSON NAMES libjsoncpp.a)
+
+find_library(LOG4CXX_LIBRARY_PATH NAMES liblog4cxx.a)
```
Comment (rdhabalia) [inline pulsar-client-cpp/CMakeLists.txt]: `liblog4cxx.a` is this correct?

**Comment by saandrews** [comment]
With this change, are we going to generate only static library? Having the ability to generate both might be useful.

**Comment by merlimat** [comment]
>  With this change, are we going to generate only static library? Having the ability to generate both might be useful.

The build always generates both `libpulsar.so` and `libpulsar.a` as before. The only difference is that `libpulsar.so` is statically linking the dependencies to have a self-contained library.

I can make the static linking configurable at the CMake level. Unfortunately it's not easy to have both builds since it needs to search for different dependencies.

**Comment by saandrews** [comment]
>I can make the static linking configurable at the CMake level. Unfortunately it's not easy to have both builds since it needs to search for different dependencies.

If it easy to get these using different targets, let us try to get that as an option. Else, we can address it later.

**Comment by merlimat** [comment]
On Thu, Jul 6, 2017 at 10:05 AM, Andrews <notifications@github.com> wrote:

> If it easy to get these using different targets, let us try to get that as
> an option. Else, we can address it later.
>
No, it won't work, because CMake either looks (and links with) for
libbboost_thread.so or for libboost.a. That's independent from the targets.


--
Matteo Merli
<matteo.merli@gmail.com>

**Comment by merlimat** [comment]
We can only have a flag that decides whether to either link statically or
dynamically with the dependencies.

--
Matteo Merli
<matteo.merli@gmail.com>

**Comment by saandrews** [comment]
>We can only have a flag that decides whether to either link statically or
dynamically with the dependencies.

This is what I was referring to.

**Comment by merlimat** [comment]
@saandrews @jai1 @rdhabalia Updated. Added flag to enable static linking. default is off.

**Pull request merged**